### PR TITLE
Feat: expand kafka broker metrics

### PR DESCRIPTION
**Pull request by jcountsNR**
Rebasing and resubmitting from the original PR due to new team ownership.

ORIGINAL PR: [23646](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23646)

### CODE REVIEW

**Comment by jcountsNR** [comment]
@dmitryax , I didn't have a way to update the other PR since I did not raise it, so unfortunately I had to create a new one. I did however rebase with main and run `make gotidy`.

For the question that was posed in the previous PR - 
'Isn't count the cumulative sum?' - New Relic views count as a tally of number of metrics rather than a sum. 
For example, if 10 metrics are sent each with a value of 10, the count would be `10`, and the sum would be `100`. Hopefully that clarifies.

**Comment by dmitryax** [comment]
> 'Isn't count the cumulative sum?' - New Relic views count as a tally of number of metrics rather than a sum.
For example, if 10 metrics are sent each with a value of 10, the count would be 10, and the sum would be 100. Hopefully that clarifies.

Ok, but this is not a New Relic specific component. Sums are usually preferable in OpenTelemetry because they are more resilient to data points misses. You can use https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/cumulativetodeltaprocessor to translate all the cumulative metrics to deltas.

Also, please take a look at metrics already defined in the Semantic Conventions https://github.com/open-telemetry/semantic-conventions/blob/ca881074f6df330bacdaa89372f2851c26545fee/docs/messaging/kafka.md?plain=1#L104C13-L104C29. `messaging.kafka.network.io` is already defined as a sum (cumulative counter) and the definition seems like the same as `kafka.brokers.incoming_byte_rate` and `kafka.brokers.outgoing_byte_rate`. We should probably suggest changing it to `messaging.kafka.brokers.network.io`.

**Comment by jcountsNR** [comment]
> > 'Isn't count the cumulative sum?' - New Relic views count as a tally of number of metrics rather than a sum.
> > For example, if 10 metrics are sent each with a value of 10, the count would be 10, and the sum would be 100. Hopefully that clarifies.
> 
> Ok, but this is not a New Relic specific component. Sums are usually preferable in OpenTelemetry because they are more resilient to data points misses. You can use https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/cumulativetodeltaprocessor to translate all the cumulative metrics to deltas.
> 
> Also, please take a look at metrics already defined in the Semantic Conventions https://github.com/open-telemetry/semantic-conventions/blob/ca881074f6df330bacdaa89372f2851c26545fee/docs/messaging/kafka.md?plain=1#L104C13-L104C29. `messaging.kafka.network.io` is already defined as a sum (cumulative counter) and the definition seems like the same as `kafka.brokers.incoming_byte_rate` and `kafka.brokers.outgoing_byte_rate`. We should probably suggest changing it to `messaging.kafka.brokers.network.io`.

Got it. I'm just catching up on this integration, but I'm thinking that `count` does work as a sum, and we might have misunderstood that originally. 

For the second point, changing the name to `messaging.kafka.brokers.network.io` makes total sense to me. Will this track both incoming and outgoing byte rate though? Because we had 2 unique counters defined here.

**Comment by dmitryax** [comment]
> Will this track both incoming and outgoing byte rate though? Because we had 2 unique counters defined here.

Yes, the metric currently defined in the spec `messaging.kafka.network.io` has `state` attribute with `in` and `out` values

**Comment by dmitryax** [comment]
There is a PR to semantic conventions originally submitted with this change https://github.com/open-telemetry/semantic-conventions/pull/113. I added a couple of comments, and hopefully we will get more attention from approvers/maintainers of that project

**Comment by jcountsNR** [comment]
> 

understood, I may have to resubmit that one as well because I don't have privileges on that fork. I see your comments though, I'll make those changes and hopefully make that easy for approvers.

Just so I'm clear, what is the order of approval on both of these? Do I need to get that one approved prior to this one or the other way around?

**Comment by dmitryax** [comment]
> Just so I'm clear, what is the order of approval on both of these? Do I need to get that one approved prior to this one or the other way around?

Ideally, yes. Otherwise, metrics submitted here are more likely to be changed later. But feel free to work on both in parallel. This PR has some CI failures that should be fixed. We also need to start adding them to the new `messaging.` namespace even if other metrics are not migrated yet.

**Comment by jcountsNR** [comment]
@dmitryax I had the same issue with other PR, I didn't have access to make changes to that fork so I had to create a new one. I did review your suggestions and I made the requested changes though, so hopefully this new one is in a more ready state.

https://github.com/open-telemetry/semantic-conventions/pull/196

**Comment by jcountsNR** [comment]
@dmitryax  commenting back on this one. I saw the request to run `make generate` in the failure logs, but in my local it's giving me an error that the `executable file is not found in $PATH `(mac). I'm not sure what file needs to be in the PATH so I'm having a hard time adding it.

**Comment by dmitryax** [comment]
> @dmitryax commenting back on this one. I saw the request to run make generate in the failure logs, but in my local it's giving me an error that the executable file is not found in $PATH (mac). I'm not sure what file needs to be in the PATH so I'm having a hard time adding it.

You don't have `make` installed? `xcode-select --install` should install all the dev tools

**Comment by jcountsNR** [comment]
> > @dmitryax commenting back on this one. I saw the request to run make generate in the failure logs, but in my local it's giving me an error that the executable file is not found in $PATH (mac). I'm not sure what file needs to be in the PATH so I'm having a hard time adding it.
> 
> You don't have `make` installed? `xcode-select --install` should install all the dev tools

I do have it installed. It starts to run but it gives an error on `mdatagen` executable.

doc.go:5: running "mdatagen": exec: "mdatagen": executable file not found in $PATH

**Comment by dmitryax** [comment]
Please try `make install-tools` first

**Comment by jcountsNR** [comment]
> Please try `make install-tools` first

Can you run the workflow again? The unit tests are passing in my local now and it looks like the lint issues are already resolved.

Also I just ran this but I end up with the same error for some reason.

**Comment by dmitryax** [comment]
Given that kafka metrics are now prefixed with the `messaging.` in the semantic conventions, I think we should adopt it for the newly introduced metrics to avoid breaking changes going forward

**Comment by jcountsNR** [comment]
@dmitryax  Agreed, I'll update them shortly. I'm working on setting this up in an EC2 instance so I can run the workflows easier. Looks like the unit test that was failing passed but now there is a different failure, so I'll try to get that resolved also.

**Comment by jcountsNR** [comment]
@dmitryax  looks like I missed one spot where the metrics still had the old name, I just updated them and reran all the unit tests. Not sure why they passed last time.

File: receiver/kafkametricsreceiver/metadata.yaml [hunk 1]
```diff
@@ -18,17 +18,125 @@ attributes:
...
+    unit: "{brokers}"
+    gauge:
+      value_type: int
+    warnings:
+      if_enabled: The metric is deprecated and will be removed. Use `messaging.kafka.broker.count`
+
+  messaging.kafka.broker.count:
+    enabled: false
     description: Number of brokers in the cluster.
     unit: "{brokers}"
```
Comment (dmitryax) [inline receiver/kafkametricsreceiver/metadata.yaml:40]: Please update all the other units to the singular form. This is a part of the recent OTel spec guidelines

```suggestion
    unit: "{broker}"
```

File: receiver/kafkametricsreceiver/metadata.yaml [hunk 2]
```diff
@@ -18,17 +18,125 @@ attributes:
...
+    description: Average consumer fetch Rate
+    unit: "{fetches}/s"
+    gauge:
+      value_type: double
+    attributes: [broker]
+    warnings:
+      if_enabled_not_set: This metric will be enabled by default in the next versions.
+  messaging.kafka.broker.incoming_byte_rate:
+    enabled: false
+    description: Average tncoming Byte Rate in bytes/second
```
Comment (dmitryax) [inline receiver/kafkametricsreceiver/metadata.yaml:58]:
```suggestion
    description: Average incoming Byte Rate in bytes/second
```

File: receiver/kafkametricsreceiver/metadata.yaml [hunk 3]
```diff
@@ -18,17 +18,125 @@ attributes:
...
+    unit: 1
```
Comment (dmitryax) [inline receiver/kafkametricsreceiver/metadata.yaml:59]:
```suggestion
    unit: By/s
```

File: receiver/kafkametricsreceiver/metadata.yaml [hunk 4]
```diff
@@ -18,17 +18,125 @@ attributes:
...
+    gauge:
+      value_type: double
+    attributes: [broker]
+    warnings:
+      if_enabled_not_set: This metric will be enabled by default in the next versions.
+  messaging.kafka.broker.request_latency:
+    enabled: false
+    description: Average request latency in ms
+    unit: "ms"
```
Comment (dmitryax) [inline receiver/kafkametricsreceiver/metadata.yaml:77]: Should be in seconds https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md#instrument-units
```suggestion
    unit: "s"
```

File: receiver/kafkametricsreceiver/metadata.yaml [hunk 5]
```diff
@@ -18,17 +18,125 @@ attributes:
...
+      value_type: int
+    attributes: [broker]
+    warnings:
+      if_enabled_not_set: This metric will be enabled by default in the next versions.
+  messaging.kafka.broker.consumer_fetch_count:
+    enabled: false
+    description: Count of consumer fetches
+    unit: "{fetches}"
+    sum:
+      value_type: double
```
Comment (dmitryax) [inline receiver/kafkametricsreceiver/metadata.yaml:133]: counters should be int

File: receiver/kafkametricsreceiver/metadata.yaml [hunk 6]
```diff
@@ -18,17 +18,125 @@ attributes:
...
+    type: int
 
 metrics:
+
   #  brokers scraper
   kafka.brokers:
     enabled: true
+    description: "[DEPRECATED] Number of brokers in the cluster."
+    unit: "{brokers}"
+    gauge:
```
Comment (dmitryax) [inline receiver/kafkametricsreceiver/metadata.yaml:32]: Please don't change the type. It should stay sum

**Review by dmitryax — COMMENTED** [review]
@jcountsNR can you please add the broker metrics to the integration test?

**Comment by jcountsNR** [comment]
> @jcountsNR can you please add the broker metrics to the integration test?

Sorry for the delay on this one. I could not get the test to run locally, so I had to run it in EC2, hence all the commits. I added the broker metrics to the integration test, and it passes when I run it, and I verified that the metrics are being collected as well by running the collector with everything enabled. The only thing I'm missing is writing to the expected.yaml, for some reason it keeps writing a blank object, but it is working so I don't know why it's doing that.

**Comment by jcountsNR** [comment]
@dmitryax , let me know if this works. I added the metrics to the test, and I also tested the collector to make sure they were coming.

File: receiver/kafkametricsreceiver/integration_test.go [hunk 7]
```diff
@@ -67,13 +67,41 @@ func TestIntegration(t *testing.T) {
...
+				rCfg.Metrics.MessagingKafkaBrokerConsumerFetchRate.Enabled = true
+				rCfg.Metrics.MessagingKafkaBrokerCount.Enabled = true
+				rCfg.Metrics.MessagingKafkaBrokerIncomingByteRate.Enabled = true
+				rCfg.Metrics.MessagingKafkaBrokerOutgoingByteRate.Enabled = true
+				rCfg.Metrics.MessagingKafkaBrokerRequestLatency.Enabled = true
+				rCfg.Metrics.MessagingKafkaBrokerRequestRate.Enabled = true
+				rCfg.Metrics.MessagingKafkaBrokerRequestSize.Enabled = true
+				rCfg.Metrics.MessagingKafkaBrokerRequestsInFlight.Enabled = true
+				rCfg.Metrics.MessagingKafkaBrokerResponseRate.Enabled = true
+				rCfg.Metrics.MessagingKafkaBrokerResponseSize.Enabled = true
```
[the remark this answers is not in the document]
Comment (jcountsNR) [reply receiver/kafkametricsreceiver/integration_test.go:96]: Hmm, I never got it to write anything to expected.yaml, but I was getting metrics from it when I ran the collector. That does seem to indicate it's an issue in the integration test, but the only thing I've changed here is just enabling the metrics.

**Pull request closed without merging**